File: src/main/java/com/joecollins/graphics/utils/StandardFont.java
package com.joecollins.graphics.utils;

import java.awt.Font;
import java.awt.FontFormatException;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

public class StandardFont {

  private StandardFont() {}

  private static final Font KLAVIKA_BOLD;
  private static final Font KLAVIKA_REGULAR;

  static {
    try {
      KLAVIKA_BOLD =
          Font.createFont(
              Font.TRUETYPE_FONT,
              StandardFont.class.getClassLoader().getResourceAsStream("Klavika Bold.otf"));
      KLAVIKA_REGULAR =
          Font.createFont(
              Font.TRUETYPE_FONT,
              StandardFont.class.getClassLoader().getResourceAsStream("Klavika Regular.otf"));
    } catch (FontFormatException | IOException e) {
      throw new RuntimeException(e);
    }
  }

  private static Map<Integer, Font> boldFontCache = new HashMap<>();
  private static Map<Integer, Font> normalFontCache = new HashMap<>();

  private static Function<Integer, Font> boldFont =
      size -> KLAVIKA_BOLD.deriveFont(Font.PLAIN, size);

  private static Function<Integer, Font> normalFont =
      size -> KLAVIKA_REGULAR.deriveFont(Font.PLAIN, size);

  public static synchronized Font readBoldFont(int size) {
    return boldFontCache.computeIfAbsent(size, boldFont);
  }

  public static synchronized Font readNormalFont(int size) {
    return normalFontCache.computeIfAbsent(size, normalFont);
  }

  public static synchronized Void setFont(String name, int style) {
    boldFont =
        size -> {
          try {
            return Font.createFont(
                    Font.TRUETYPE_FONT,
                    StandardFont.class.getClassLoader().getResourceAsStream(name))
                .deriveFont(Font.BOLD | style, size);
          } catch (FontFormatException | IOException e) {
            throw new RuntimeException(e);
          }
        };
    normalFont =
        size -> {
          try {
            return Font.createFont(
                    Font.TRUETYPE_FONT,
                    StandardFont.class.getClassLoader().getResourceAsStream(name))
                .deriveFont(style, size);
          } catch (FontFormatException | IOException e) {
            throw new RuntimeException(e);
          }
        };
    boldFontCache.clear();
    normalFontCache.clear();
    return null;
  }
}

File: src/main/java/com/joecollins/graphics/utils/StandardFont.kt
package com.joecollins.graphics.utils

import java.awt.Font
import java.awt.FontFormatException
import java.io.IOException
import java.lang.RuntimeException
import java.lang.Void
import java.util.HashMap

object StandardFont {
    private val KLAVIKA_BOLD: Font
    private val KLAVIKA_REGULAR: Font

    init {
        try {
            KLAVIKA_BOLD = Font.createFont(
                    Font.TRUETYPE_FONT,
                    StandardFont::class.java.classLoader.getResourceAsStream("Klavika Bold.otf"))
            KLAVIKA_REGULAR = Font.createFont(
                    Font.TRUETYPE_FONT,
                    StandardFont::class.java.classLoader.getResourceAsStream("Klavika Regular.otf"))
        } catch (e: FontFormatException) {
            throw RuntimeException(e)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    private val boldFontCache: MutableMap<Int, Font> = HashMap()
    private val normalFontCache: MutableMap<Int, Font> = HashMap()
    private var boldFont = { size: Int -> KLAVIKA_BOLD.deriveFont(Font.PLAIN, size.toFloat()) }
    private var normalFont = { size: Int -> KLAVIKA_REGULAR.deriveFont(Font.PLAIN, size.toFloat()) }

    @Synchronized
    @JvmStatic fun readBoldFont(size: Int): Font {
        return boldFontCache.computeIfAbsent(size, boldFont)
    }

    @Synchronized
    @JvmStatic fun readNormalFont(size: Int): Font {
        return normalFontCache.computeIfAbsent(size, normalFont)
    }

    @Synchronized
    @JvmStatic fun setFont(name: String, style: Int): Void? {
        boldFont = { size ->
            try {
                Font.createFont(
                        Font.TRUETYPE_FONT,
                        StandardFont::class.java.classLoader.getResourceAsStream(name))
                        .deriveFont(Font.BOLD or style, size.toFloat())
            } catch (e: FontFormatException) {
                throw RuntimeException(e)
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
        normalFont = { size ->
            try {
                Font.createFont(
                        Font.TRUETYPE_FONT,
                        StandardFont::class.java.classLoader.getResourceAsStream(name))
                        .deriveFont(style, size.toFloat())
            } catch (e: FontFormatException) {
                throw RuntimeException(e)
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
        boldFontCache.clear()
        normalFontCache.clear()
        return null
    }
}
